pipeline-runner/tests/testthat/test-qc-4-filter_gene_umi_outlier.R

```diff
@@ -0,0 +1,87 @@
+mock_config <- function() {
+    config <- list(
+        auto = TRUE,
+        enabled = TRUE,
+        filterSettings = list(
+            regressionType = 'gam',
+            regressionTypeSettings = list(
+                gam = list(p.level = 0.1)
+            )
+        )
+    )
+
+    return(config)
+}
+
+
+mock_scdata <- function() {
+    pbmc_raw <- read.table(
+        file = system.file("extdata", "pbmc_raw.txt", package = "Seurat"),
+        as.is = TRUE)
+
+    scdata <- Seurat::CreateSeuratObject(counts = pbmc_raw)
+
+    # add samples
+    scdata$samples <- rep(c('123abc', '123def'), each = 40)
+    scdata <- Seurat::RenameCells(scdata, paste(scdata$samples, colnames(scdata), sep = ''))
+    return(scdata)
+}
+
+
+test_that("filter_gene_umi_outlier updates gam p.level in config if auto", {
+    scdata <- mock_scdata()
+    config <- mock_config()
+    config$filterSettings$regressionTypeSettings$gam$p.level <- 1
+    out <- filter_gene_umi_outlier(scdata, config, '123def')
+    new <- out$config$filterSettings$regressionTypeSettings$gam$p.level
+
+    expect_lt(new, 1)
+})
+
+test_that("filter_gene_umi_outlier can filter cells", {
+    scdata <- mock_scdata()
+    config <- mock_config()
+    nstart <- ncol(scdata)
+
+    out <- filter_gene_umi_outlier(scdata, config, '123def')
+    expect_lt(ncol(out$data), nstart)
+})
+
+test_that("filter_gene_umi_outlier is sample specific", {
+    scdata <- mock_scdata()
+    config <- mock_config()
+
+    out1 <- filter_gene_umi_outlier(scdata, config, '123abc')
+    out2 <- filter_gene_umi_outlier(scdata, config, '123def')
+    expect_lt(ncol(out2$data), ncol(out1$data))
```

Comment:
All the pr seems good, my only comment is why we would expect to see less ncol in out2 than out1 and why that ensures that the filter is sample-specific

Reply:
it just happens that in this example dataset it is the case that the only cells filtered are in the last 40 cells i.e. sample `'123def'`. This just makes sure that we don't acidentally break the sample subsetting in the filter (if e.g. we ended up applying it to all samples at the same time then the results would be equal).

Reply:
Cool thanks. Feel free to merge